Add support for indexed functions

daniel-raffler opened the pull request:
Hello,
this PR will add support for indexed functions to the JavaSMT API. In indexed functions take additional arguments that are not terms, but have to be constant. One example is `(_ extract i j)` for bitvector, which takes two more parameters `i` and `j` to specify the range of bits that should be returned. While `extract` is available in `BitvectorFormulaManager` and we can create the term in JavaSMT, we've so far lacked a way to access the indices in the visitor.

In this PR we will address the problem by adding a new method `ImmutableList<Integer> getIndices()` to the `FunctionDeclaration` API and implementing it for all solvers. In the `FormulaVisitor` the method can then be used to get the values of the parameters from the function symbol:
```java
new FormulaVisitor<Formula>() {
  @Override
    public Formula visitFunction(
      Formula f,
      List<Formula> args,
      FunctionDeclaration<?> functionDeclaration) {
    List<Integer> params = functionDeclaration.getIndices();
    if (functionDeclaration.getKind().equals(FunctionDeclarationKind.BV_EXTRACT)) {
      // Use the indices to recreate the formula
      return bvmgr.extract((BitvectorFormula) args.get(0), params.get(0), params.get(1));
    }
  }
}
```

**Open issues**
- Fix the implementation for Princess

daniel-raffler committed:
Indices: Extend the API to support indexed function symbols, such as like "extract" for bitvectors

daniel-raffler committed:
Indices: Add support for indexed functions symbols to the solver backends for Bitwuzla, CVC5, MathSAT, Princess and Z3

kfriedberger reviewed (commented):
My first thought was that we can actually model all indices as plain IntegerFormula-based arguments and just revert this encoding on usage.  Seperating the concept of arguments into "args and indices" while still having methods that allow both, seems weird.

For example, `BV_ROTATE_LEFT` exists in two versions, one with integer-based shift (modeled as index), one with IntegerFormula-based shift (modeled as argument).

Please give some more reason to model indices separatelly.

daniel-raffler committed:
Merge branch 'master' into indexed_functions

# Conflicts:
#	src/org/sosy_lab/java_smt/solvers/princess/PrincessFormulaCreator.java

daniel-raffler committed:
Fix merge

daniel-raffler committed:
Indices: Fix handling of bv_extract and bv_concat in Princess

daniel-raffler committed:
Indices: Fix CVC5 visitor for sep.nil

daniel-raffler committed:
Merge branch 'refs/heads/master' into indexed_functions

# Conflicts:
#	src/org/sosy_lab/java_smt/solvers/princess/PrincessEnvironment.java

daniel-raffler committed:
Indices: Add tests for visiting indexed functions

daniel-raffler committed:
Indices: Add support for new solver "Z3 with interpolation"

daniel-raffler committed:
Indices: Fix indices in CVC5 visitor

We need to check `hasOp` before getting the operator of a term to extract the indices. Otherwise, it will crash for SEP_EMP, which has no operator

daniel-raffler committed:
Merge branch 'refs/heads/master' into indexed_functions

daniel-raffler committed:
Indices: Add support for indexed functions in the tracer

daniel-raffler committed:
Indices: Add support for indices in Yices, and update the tests

daniel-raffler committed:
Merge branch 'refs/heads/master' into indexed_functions

# Conflicts:
#	src/org/sosy_lab/java_smt/basicimpl/FunctionDeclarationImpl.java
#	src/org/sosy_lab/java_smt/delegate/trace/TraceFormulaManager.java
#	src/org/sosy_lab/java_smt/solvers/princess/PrincessFormulaCreator.java

daniel-raffler committed:
Indices: Remove duplicate code

daniel-raffler committed:
Apply refaster patch

daniel-raffler committed:
Merge branch 'refs/heads/master' into indexed_functions

daniel-raffler committed:
Merge branch 'master' into indexed_functions

daniel-raffler committed:
Merge branch 'master' into indexed_functions
